### PR TITLE
asmrefine pfx (prefix for finding constants via name) + Kernel_C tweaks

### PR DESCRIPTION
**Pull request by Xaphiosis**
@mbrcknl : hoping this doesn't break other BV efforts

Needs full regression test to make sure I didn't derp anything else.

### CODE REVIEW

File: tools/asmrefine/SimplExport.thy [hunk 1]
```diff
@@ -1105,15 +1110,18 @@ fun emit_func_body ctxt outfile eparams name = let
     else (emit outfile ("1 Basic Ret 0");
           emit_body ctxt outfile eparams body 2 "1" ("ErrExc", "ErrExc")
             |> snd |> prefix "EntryPoint " |> emit outfile
+(*
           handle TERM (s, ts) => raise TERM ("emit_func_body: " ^ name ^ ": " ^ s, body :: @{term True} :: ts)
             | TYPE (s, Ts, ts) => raise TYPE ("emit_func_body: " ^ name ^ ": " ^ s, Ts, body :: @{term True} :: ts)
             | Empty => raise TERM ("emit_func_body: Empty", [body]))
+*)
+)
```
Comment (mbrcknl) [inline tools/asmrefine/SimplExport.thy:1118]: Did you mean to leave this commented out?

File: tools/asmrefine/SimplExport.thy [hunk 2]
```diff
@@ -1063,10 +1067,11 @@ fun emit_body ctxt outfile params (Const (@{const_name Seq}, _) $ a $ b) n c e =
   | emit_body _ _ _ t _ _ _ = raise TERM ("emit_body", [t])
 
 fun emit_func_body ctxt outfile eparams name = let
+    val _ = warning ("EMIT_FUNC_BODY " ^ name)
```
Comment (mbrcknl) [inline tools/asmrefine/SimplExport.thy:1070]: Did you mean to leave this tracing here?

**Review by mbrcknl — APPROVED** [review]
I've not kept up with the Kernel_Config development, so I don't fully grok everything here. But I think I understand enough to see roughly why the `pfx` is needed in asmrefine.

In any case, I think this should be fine for the rest of BV. If it makes it through SimplExportAndRefine, then it shouldn't affect anything downstream.

File: tools/asmrefine/SimplExport.thy (hunk 1 above)
Comment (Xaphiosis) [reply tools/asmrefine/SimplExport.thy:1118]: Embarrassing medical incident in the git surgery (and the tracing too), thank you for spotting! Will fix.

Comment (Xaphiosis) [reply tools/asmrefine/SimplExport.thy:1118]: Though I'll wait until regression proof tests complete, as these don't change the success criteria.

**Comment by Xaphiosis** [comment]
Oh great, AsmRefine failed on the ARMs. Needs work.

**Comment by mbrcknl** [comment]
> Oh great, AsmRefine failed on the ARMs. Needs work.

Probably not anything special about the ARMs. Apparently, we never turned on AsmRefine for RISC-V. The contents of the AsmRefine session get imported by SimplExportAndRefine, so the only thing that's not checked on RISC-V is AsmRefineTest. Of course, that's what failed on the ARMs.

I don't understand the error, though...

```
*** exception Io {cause = SysErr ("No such file or directory", SOME ENOENT),
***               function = "TextIO.openIn", name =
***               "/home/test-runner/ver/l4v/tools/asmrefine/testfiles/global_array_swap_Cfuns.txt"} raised (line 284 of "./basis/TextIO.sml")
*** At command "ML" (line 114 of "~/ver/l4v/tools/asmrefine/testfiles/global_array_swap_gref.thy")
```

I can't see how you get to global_array_swap_gref.thy:114 without creating the file at line 28 (`emit_C_everything_relative`), assuming that the `@{theory}` at line 114 is the same as the theory of the `@{context}` at line 28.

**Comment by Xaphiosis** [comment]
> I can't see how you get to global_array_swap_gref.thy:114 without creating the file at line 28 (`emit_C_everything_relative`), assuming that the `@{theory}` at line 114 is the same as the theory of the `@{context}` at line 28.

This is a derp by me. I added an extra argument `pfx` but then I don't pass anything for it for `emit_C_everything_relative` in the `asmrefine/testfiles`, so ML happily hands me back a partial function and doesn't generate any files, hence the IO Error.

**Comment by Xaphiosis** [comment]
It's still somewhat strange that we test AsmRefine on ARM and ARM_HYP, but SimplExportAndRefine on ARM and RISCV64... should we have turned that on for RISCV64?

**Comment by lsf37** [comment]
> @mbrcknl :
> It's still somewhat strange that we test AsmRefine on ARM and ARM_HYP, but SimplExportAndRefine on ARM and RISCV64... should we have turned that on for RISCV64?

I think it would make sense to turn it on for RISCV64 (at least if you can get it to work without putting in a lot of time)

**Comment by Xaphiosis** [comment]
> > @mbrcknl :
> > It's still somewhat strange that we test AsmRefine on ARM and ARM_HYP, but SimplExportAndRefine on ARM and RISCV64... should we have turned that on for RISCV64?
> 
> I think it would make sense to turn it on for RISCV64 (at least if you can get it to work without putting in a lot of time)

Ok I'll give it a go.

**Comment by mbrcknl** [comment]
> > I can't see how you get to global_array_swap_gref.thy:114 without creating the file at line 28 (`emit_C_everything_relative`), assuming that the `@{theory}` at line 114 is the same as the theory of the `@{context}` at line 28.
> 
> This is a derp by me. I added an extra argument `pfx` but then I don't pass anything for it for `emit_C_everything_relative` in the `asmrefine/testfiles`, so ML happily hands me back a partial function and doesn't generate any files, hence the IO Error.

Ah yes, I didn't make that connection. :-)

> @mbrcknl :
> It's still somewhat strange that we test AsmRefine on ARM and ARM_HYP, but SimplExportAndRefine on ARM and RISCV64... should we have turned that on for RISCV64?

This is probably just an oversight (by me) when we set up SimplExportAndRefine on RISC-V. We should have also turned on AsmRefine.

Turning AsmRefine on for ARM_HYP was probably just because it's cheap and we could, but we didn't turn on SimplExportAndRefine because it's expensive.

**Comment by Xaphiosis** [comment]
Yeah, I can see why it wasn't turned on, it needs some kind of per-arch testfiles folder thing because of `global_asm_stmt.c` which uses `int` to do an inline asm register read, which yields a nasty looking error:
```
exception TYPE raised (line 474 of "~/repos/v2/l4v/tools/c-parser/isar_install.ML"):
  type_of: type mismatch in application
  32 word
  64 word
  (λx _. x) B.1
```
Which is fair enough, using an ARM test file with inline assembly isn't going to work on RISCV64. To get this to work, we'd need separate arch folders and an inline asm example for each. I might leave that to another pull request.

**Pull request merged**